### PR TITLE
Add an Offset field to WithDamageOverlayInfo

### LINKED ISSUE #21113: Add Offset for WithDamageOverlay trait
MOTIVATION

Together with #20769 it allow to recreate fire and smoke VFX for damaged buildings.

ALTERNATIVES

or add Duration for WithIdleOverlay
or add timed condition into ORA :)

### PR DESCRIPTION
**Pull request by abcdefg30**
Closes #21113.

### CODE REVIEW

**Comment by PunkPun** [comment]
Shouldn't the offset really be random?

**Comment by PunkPun** [comment]
Something like #19238

**Comment by Porenutak** [comment]
In D2k  offset is not random, but the duration is (or number of loops). Also when HP goes under 50% not all animation start at once, but looks like animation start have some randomness too. 
But I dont think its necessary to be exact the same as in original. Its just a decoration. With Offset and number of loops we can get close enough to look good.

**Comment by abcdefg30** [comment]
> Shouldn't the offset really be random?

Unless there is a specific use-case requested I wouldn't make this more complicated than necessary. If anything the trait should be made conditional and then you can emulate the behavior even better by using multiple instances.

File: OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@
 namespace OpenRA.Mods.Common.Traits.Render
 {
 	[Desc("Renders an overlay when the actor is taking heavy damage.")]
-	public class WithDamageOverlayInfo : TraitInfo, Requires<RenderSpritesInfo>
+	public class WithDamageOverlayInfo : TraitInfo, Requires<RenderSpritesInfo>, Requires<BodyOrientationInfo>
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs]: shouldn't body orientation be optional?

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs]: We can make it optional, but that trait does not make sense without it. (All our actors have body orientation.) `WithIdleOverlay` also requires it.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs]: Adjusted it.

File: OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs [hunk 2]
```diff
@@ -16,7 +16,7 @@
 namespace OpenRA.Mods.Common.Traits.Render
 {
 	[Desc("Renders an overlay when the actor is taking heavy damage.")]
-	public class WithDamageOverlayInfo : TraitInfo, Requires<RenderSpritesInfo>
+	public class WithDamageOverlayInfo : TraitInfo, Requires<RenderSpritesInfo>, NotBefore<BodyOrientationInfo>, IRulesetLoaded
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs]: the proper way we check is by delaying the acquisition of the trait to INotifyCreated

Comment (michaeldgg2) [reply OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs]: Or using `Requires`? That way the check in `RulesetLoaded` is redundant. But yes, if this trait doesn't absolutely require a trait, it should use `INotifyCreated` to retrieve a reference to it.

**Comment by abcdefg30** [comment]
Updated.

**Review by PunkPun — APPROVED** [review]
LGTM

**Comment by PunkPun** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e15ac7f47b7e6597fc86a722003a9db6bae6ff76...604bade6590795b984c02c76803078c4dbaf8592)

**Pull request merged**